operators/src/pro/cache/cache_tiles.rs

```diff
@@ -0,0 +1,336 @@
+            }
+            CachedTiles::U64(v) => {
+                TypedCacheTileStream::U64(CacheTileStream::new(v.clone(), *query))
+            }
+            CachedTiles::I8(v) => TypedCacheTileStream::I8(CacheTileStream::new(v.clone(), *query)),
+            CachedTiles::I16(v) => {
+                TypedCacheTileStream::I16(CacheTileStream::new(v.clone(), *query))
+            }
+            CachedTiles::I32(v) => {
+                TypedCacheTileStream::I32(CacheTileStream::new(v.clone(), *query))
+            }
+            CachedTiles::I64(v) => {
+                TypedCacheTileStream::I64(CacheTileStream::new(v.clone(), *query))
+            }
+            CachedTiles::F32(v) => {
+                TypedCacheTileStream::F32(CacheTileStream::new(v.clone(), *query))
+            }
+            CachedTiles::F64(v) => {
+                TypedCacheTileStream::F64(CacheTileStream::new(v.clone(), *query))
+            }
+        }
+    }
+}
+
+impl CacheElementsContainerInfos<RasterQueryRectangle> for CachedTiles {
+    fn is_expired(&self) -> bool {
+        self.is_expired()
+    }
+}
+
+impl<T> CacheElementsContainer<RasterQueryRectangle, RasterTile2D<T>> for CachedTiles
+where
+    T: CacheElementSubType<CacheElementType = RasterTile2D<T>> + Pixel,
+{
+    type ResultStream = CacheTileStream<T>;
+
+    fn result_stream(&self, query: &RasterQueryRectangle) -> Option<CacheTileStream<T>> {
+        T::result_stream(self, query)
+    }
+}
+
+impl<T> LandingZoneElementsContainer<RasterTile2D<T>> for LandingZoneQueryTiles
+where
+    T: CacheElementSubType<CacheElementType = RasterTile2D<T>> + Pixel,
+{
+    fn insert_element(&mut self, element: RasterTile2D<T>) -> Result<(), super::error::CacheError> {
+        T::insert_element_into_landing_zone(self, element)
+    }
+
+    fn create_empty() -> Self {
+        T::create_empty_landing_zone()
+    }
+}
+
+impl<T> CacheElement for RasterTile2D<T>
+where
+    T: Pixel + CacheElementSubType<CacheElementType = Self>,
+{
+    type Query = RasterQueryRectangle;
+    type LandingZoneContainer = LandingZoneQueryTiles;
+    type CacheContainer = CachedTiles;
+    type ResultStream = CacheTileStream<T>;
+    type CacheElementSubType = T;
+
+    fn cache_hint(&self) -> geoengine_datatypes::primitives::CacheHint {
+        self.cache_hint
+    }
+
+    fn typed_canonical_operator_name(
+        key: crate::engine::CanonicOperatorName,
+    ) -> super::shared_cache::TypedCanonicOperatorName {
+        super::shared_cache::TypedCanonicOperatorName::Raster(key)
+    }
+
+    fn update_stored_query(&self, query: &mut Self::Query) -> Result<(), CacheError> {
+        query.spatial_bounds.extend(&self.spatial_partition());
+        query.time_interval = query
+            .time_interval
+            .union(&self.time)
+            .map_err(|_| CacheError::ElementAndQueryDoNotIntersect)?;
+        Ok(())
+    }
+}
+
+macro_rules! impl_cache_element_subtype_magic {
```

Comment:
what is the magic here?

Reply:
5950e70ee95cb512487d70a84f0467cadb1b0e9c

Reply:
[5950e70](https://github.com/geo-engine/geoengine/pull/833/commits/5950e70ee95cb512487d70a84f0467cadb1b0e9c)